crates/energy/src/graph.rs

```diff
@@ -0,0 +1,259 @@
+use std::collections::HashSet;
+
+use bevy::prelude::*;
+use bevy::utils::petgraph::prelude::*;
+#[cfg(feature = "energy_graph_debug_lines")]
+use bevy::utils::petgraph::visit::IntoNodeReferences;
+#[cfg(feature = "energy_graph_debug_lines")]
+use bevy_prototype_debug_lines::{DebugLines, DebugLinesPlugin};
+use de_core::gamestate::GameState;
+use de_core::objects::Active;
+use de_core::projection::ToFlat;
+use de_index::SpatialQuery;
+use de_spawner::{DespawnEventsPlugin, DespawnedComponentsEvent};
+use parry3d::bounding_volume::Aabb;
+use parry3d::math::Point;
+use tinyvec::TinyVec;
+
+// The max distance (in meters) between two entities for them to be consider neighbors in the graph
+const MAX_DISTANCE: f32 = 10.0;
+
+pub(crate) struct GraphPlugin;
+
+impl Plugin for GraphPlugin {
+    fn build(&self, app: &mut App) {
+        app.add_plugins(DespawnEventsPlugin::<&NearbyUnits, NearbyUnits>::default())
+            .add_systems(OnEnter(GameState::Playing), setup)
+            .add_systems(OnExit(GameState::Playing), clean_up)
+            .add_systems(PostUpdate, spawn_graph_components)
+            .add_systems(
+                FixedUpdate,
+                (
+                    remove_old_nodes.before(GraphSystemSet::UpdateNearby),
+                    update_nearby_recv.in_set(GraphSystemSet::UpdateNearby),
+                    update_graph
+                        .in_set(GraphSystemSet::UpdateGraph)
+                        .after(GraphSystemSet::UpdateNearby),
+                )
+                    .run_if(in_state(GameState::Playing)),
```

Comment:
It should run whole `AppState::InGame` because game loading phase when map entities are spawned make take longer than a single frame (especially during multiplayer).